File: kenni-client-flutter/lib/src/jwt.dart
import 'dart:convert';

/// Minimal base64url JWT payload decoder. Display-only — token signature
/// verification happens server-side at the token endpoint as part of the
/// OIDC code exchange (flutter_appauth performs it).
Map<String, dynamic> decodeJwtPayload(String token) {
  final parts = token.split('.');
  if (parts.length < 2) {
    throw const FormatException('Malformed JWT: missing payload segment');
  }
  final padded = base64Url.normalize(parts[1]);
  final json = utf8.decode(base64Url.decode(padded));
  return jsonDecode(json) as Map<String, dynamic>;
}

File: kenni-client-expo/src/jwt.ts
// Minimal base64url JWT payload decoder. We only use it to display the
// id_token claims — token signature verification happens server-side at
// the token endpoint as part of the OIDC code exchange.
export const decodeJwtPayload = (token: string): Record<string, unknown> => {
  const segment = token.split(".")[1];
  if (!segment) throw new Error("Malformed JWT: missing payload segment");
  const padded = segment + "=".repeat((4 - (segment.length % 4)) % 4);
  const base64 = padded.replace(/-/g, "+").replace(/_/g, "/");
  const json =
    typeof atob === "function"
      ? atob(base64)
      : Buffer.from(base64, "base64").toString("utf8");
  return JSON.parse(json) as Record<string, unknown>;
};
